docs/opentracing.rst

```diff
@@ -0,0 +1,179 @@
+https://opentracing.io/docs/overview/what-is-tracing/
+
+Python specific tracing concepts are at https://opentracing.io/guides/python/.
+Note that synapse wraps opentracing in a small library in order to make the
+opentracing dependency optional. That means that the access patterns are
+different to those demonstrated here. However, it is still usefull to know.
+Especially if opentracing is included as a full dependency in the future or if
+you are modifying synapse's opentracing lib.
+
+For more information about Jaeger's implementation see
+https://www.jaegertracing.io/docs/
+
+=================
+Setup opentracing
+=================
+
+To receive opentracing spans start up a Jaeger server using docker like so
+
+.. code-block:: bash
+
+   docker run -d --name jaeger \ -e COLLECTOR_ZIPKIN_HTTP_PORT=9411 \
+     -p 5775:5775/udp \
+     -p 6831:6831/udp \
+     -p 6832:6832/udp \
+     -p 5778:5778 \
+     -p 16686:16686 \
+     -p 14268:14268 \
+     -p 9411:9411 \
+     jaegertracing/all-in-one:1.13
+
+Latest documentation is probably at
+https://www.jaegertracing.io/docs/1.13/getting-started/
+
+
+Enable opentracing in synapse
+-----------------------------
+
+Opentracing is not enabled by default. It must be enabled in the homeserver
+config by uncommenting the config options under ``opentracing``. For example:
+
+.. code-block:: yaml
+
+  opentracing:
+    # Enable / disable tracer
+    tracer_enabled: true
+    # The list of homeservers we wish to expose our current traces to.
+    # The list is a list of regexes which are matched against the
+    # servername of the homeserver
+    homeserver_whitelist:
+      - ".*"
+
+Homeserver whitelisting
+-----------------------
+
+The homeserver whitelist is configured using regex. A list of regexes can be
+given and their union will be compared when propagating any spans through a
+carrier. Most of the whitelist checks are encapsulated in the lib's injection
+and extraction method but be aware that using custom carriers or crossing
+unchartered waters will require the enforcement of this whitelist.
+
+``logging/opentracing.py`` has a ``whitelisted_homeserver`` method which takes
+in a destination and compares it to the whitelist.
+
+============================
+Using opentracing in synapse
+============================
+
+Access to the opentracing api is mediated through the
+``logging/opentracing.py`` lib. Opentracing is encapsulated such that
+no statefull spans from opentracing are used in synapses code. This allows
+opentracing to be easily disabled in synapse and thereby have opentracing as
+an optional dependency. This does however limit the number of modifyable spans
+at any point in the code to one. From here out references to opentracing refer
+to the lib implemented in synapse.
+
+Tracing
+-------
+
+In synapse it is not possible to start a non-active span. Spans can be started
+using the ``opentracing.start_active_span`` method. This returns a scope (see
```

Comment:
this is probably clearer *without* the `opentracing.`, to distinguish from the real opentracing library.